### PR TITLE
DataNotAvailableLayer fixes

### PR DESCRIPTION
**Pull request by patrick-wilken**
1. Fixed one place where trying to access the size placeholder fails in `get_out_data_from_opts()` when the input is a `DataNotAvailableLayer`. I suspect there might be more such cases in the code.

2. `DataNotAvailableLayer` has to do "register_as_extern_data" because the data template might be needed at inference time, e.g. as the target of a ChoiceLayer. Usually this happens in `LayerBase.post_init()`, however this is not called for `DataNotAvailableLayer`.

### CODE REVIEW

File: returnn/tf/layers/base.py [hunk 1]
```diff
@@ -1478,6 +1478,11 @@ def __init__(self, layer_class, layer_desc, **kwargs):
     self.layer_class_ = layer_class
     self.layer_desc = layer_desc
 
+    register_as_extern_data = layer_desc.get("register_as_extern_data")
```
Comment (JackTemaki) [inline returnn/tf/layers/base.py:1481]: maybe add a comment here then (or in the class docstring) that it needs to manage the `register_as_extern_data`

**Review by JackTemaki — APPROVED** [review]
This actually solves a problem in one of the setups I avoided by explicitly giving `n_out` for output layers relying on non existent data, so thanks for finding that.

File: returnn/tf/layers/basic.py [hunk 2]
```diff
@@ -3686,7 +3686,8 @@ def map_axis_name(s):
         setattr(out, s, i)
         if s == "feature_dim_axis":
           out.dim = out.batch_shape[out.feature_dim_axis]
-    if size_base:
+    if out.size_placeholder is not None and size_base:  # size_placeholder is not set if source is DataNotAvailableLayer
```
Comment (albertz) [inline returnn/tf/layers/basic.py:3689]: I don't quite understand this comment. Why is this only for DataNotAvailableLayer, and maybe not in other cases? And why is that way at all? Shouldn't we maybe fix this in DataNotAvailableLayer? This sounds like you do some workaround here.

**Comment by albertz** [comment]
Do not explain things here in a comment in the PR, but directly in the code.

File: returnn/tf/layers/base.py [hunk 3]
```diff
@@ -1478,6 +1478,11 @@ def __init__(self, layer_class, layer_desc, **kwargs):
...
+    if register_as_extern_data:
+      self.network.extern_data.extra_added_keys.add(register_as_extern_data)
+      self.network.extern_data.data[register_as_extern_data] = self.output
```
Comment (albertz) [inline returnn/tf/layers/base.py:1484]: I don't understand. Why is this needed here?

Comment (albertz) [reply returnn/tf/layers/base.py:1484]: > Usually this happens in `LayerBase.post_init()`, however this is not called for DataNotAvailableLayer.

Why? And shouldn't we fix that? This again sounds like you do a workaround here.

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (JackTemaki) [reply returnn/tf/layers/basic.py:3689]: I think the `size_placeholder` being `None` is a valid case. This can not be fixed in the DataNotAvailableLayer, because the (correct) content for the size_placeholder does not exist, and the code would fail in any way. Introducing an arbitrary filled size_placeholder does not sound like a valid solution.

Comment (albertz) [reply returnn/tf/layers/basic.py:3689]: But then this comment is wrong (misleading), as this is not about DataNotAvailableLayer.

Comment (albertz) [reply returnn/tf/layers/basic.py:3689]: Also, as I always say, we should never just accept it as it is. We should always ask: Does this makes sense? Should it really be this way? Does it make sense that `size_placeholder` can be `None`?

File: returnn/tf/layers/base.py (hunk 3 above)
Comment (JackTemaki) [reply returnn/tf/layers/base.py:1484]: @patrick-wilken Is there any unwanted side effect in calling `post_init()` on the `DataNotAvailableLayer`? I am not sure what happens if we call `self.batch_norm` on an output which has no defined placeholder. if there are no side effects, we could call `post_init` before returning the `DataNotAvailableLayer`, if there are, this seems to be the correct solution.

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (JackTemaki) [reply returnn/tf/layers/basic.py:3689]: But it does make sense. What should the `size_placeholder` be if the data does not exist?  I think `None` is valid here. Of course you could add a `size_placeholder` with fake entries, but I do not think this is a good idea. Any variable should have a reasonable content, or not?

The comment is also correct, because as long as we are not using `DataNotAvailableLayer`, the size_placeholder should be correctly defined.

File: returnn/tf/layers/base.py (hunk 3 above)
Comment (albertz) [reply returnn/tf/layers/base.py:1484]: If `LayerBase.post_init()` does not work on `DataNotAvailableLayer`, then why do another workaround at a different place? Why not just define `DataNotAvailableLayer.post_init()`? Or fix `LayerBase.post_init()`?

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (albertz) [reply returnn/tf/layers/basic.py:3689]: But there might be other layers as well. It would be wrong to say this is only for one specific layer. The `LayerBase` class is the base and should never have specific code for specific sub instances.

Either we say `size_placeholder` can be `None` in general, or not. But saying it can only be for one specific layer is very hacky and will surely lead to other problems.

What you could do (much better):
```suggestion
    if out.size_placeholder is not None and size_base:  # size_placeholder could be unset eg. with DataNotAvailableLayer
```

**Comment by patrick-wilken** [comment]
Ok, thanks, this should be cleaner now.

File: returnn/tf/layers/base.py (hunk 1 above)
Comment (patrick-wilken) [reply returnn/tf/layers/base.py]: `register_as_extern_data` is already mentioned in the doc string of `DataNotAvailableLayer`, check if a comment is still necessary in the current code.

File: returnn/tf/network.py [hunk 4]
```diff
@@ -949,14 +949,18 @@ def _create_layer(self, name, layer_class, **layer_desc):
...
             name=layer_desc['name'], network=layer_desc['network'], output=output_template,
-            layer_class=layer_class, layer_desc=layer_desc)
-        layer = layer_class(**layer_desc)
+            layer_class=layer_class, layer_desc=layer_desc,
+            register_as_extern_data=layer_desc.get('register_as_extern_data'))
+        else:
+          layer = layer_class(**layer_desc)
         layer.post_init(layer_desc)
-        layer.output.sanity_check()
+        layer.output.sanity_check(ignore_placeholder=not placeholder_is_available)
```
Comment (albertz) [inline returnn/tf/network.py:963]: Why is this needed? If it is not set, `sanity_check` will also work.

File: returnn/tf/network.py [hunk 5]
```diff
@@ -949,14 +949,18 @@ def _create_layer(self, name, layer_class, **layer_desc):
           print("layer %s/%r output: %r" % (self.name, name, output_template))
         output_template.sanity_check(ignore_placeholder=True)  # placeholder might be overwritten later
         output_template_special_axes = output_template.get_special_axes_dict()
+        placeholder_is_available = True
```
Comment (albertz) [inline returnn/tf/network.py:952]: Why do we need this variable? Can't we just check for `layer.output.placeholder is not None` below?

**Comment by JackTemaki** [comment]
Isn't this now way more complicated than before? The change in `post_init` is good, but I would still return the `DataNotAvailableLayer` directly where it is defined, just applying post_init at that point.

File: returnn/tf/network.py (hunk 4 above)
Comment (patrick-wilken) [reply returnn/tf/network.py]: Oh, didn't notice.

File: returnn/tf/network.py (hunk 5 above)
Comment (patrick-wilken) [reply returnn/tf/network.py]: Fixed now.

File: returnn/tf/network.py [hunk 6]
```diff
@@ -972,22 +974,23 @@ def _create_layer(self, name, layer_class, **layer_desc):
...
         from returnn.tf.util.basic import identity_with_check_numerics
         layer.output.placeholder = identity_with_check_numerics(
           layer.output.placeholder,
           name="%s_identity_with_check_numerics_output" % layer_class.cls_get_tf_scope_name(name))
     assert layer.output
-    assert layer.output.placeholder is not None
-    layer.output.placeholder.set_shape(layer.output.batch_shape)
-    assert layer.output.size_placeholder is not None
+    if layer.output.placeholder is not None:
+      assert layer.output.placeholder is not None
```
Comment (patrick-wilken) [inline returnn/tf/network.py:991]: Oops, let me remove this line :D

**Review by albertz — APPROVED** [review]
Ok looks fine now. Let's wait for the tests.
